libopae++/src/properties.cpp

```diff
@@ -69,6 +69,279 @@ properties::properties()
+      capabilities.invalidate();
+      num_mmio.invalidate();
+      num_interrupts.invalidate();
+      accelerator_state.invalidate();
+      object_id.invalidate();
+      parent.invalidate();
+      guid.invalidate();
+    }
+
+    if (p.props_) {
+      if ((res = fpgaCloneProperties(p.props_, &props_)) != FPGA_OK) {
+        log_.error() << "fpgaCloneProperties() failed with (" << res
+                     << ") " << fpgaErrStr(res);
+        throw except(res, OPAECXX_HERE);
+      }
+
+      // make sure that we have a cached copy of
+      // everything that p has cached.
+      if (p.type.is_set()) {
+        res = type.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.bus.is_set()) {
+        res = bus.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.device.is_set()) {
+        res = device.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.function.is_set()) {
+        res = function.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.socket_id.is_set()) {
+        res = socket_id.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.num_slots.is_set()) {
+        res = num_slots.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.bbs_id.is_set()) {
+        res = bbs_id.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.bbs_version.is_set()) {
+        res = bbs_version.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.vendor_id.is_set()) {
+        res = vendor_id.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.model.is_set()) {
+        res = model.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.local_memory_size.is_set()) {
+        res = local_memory_size.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.capabilities.is_set()) {
+        res = capabilities.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.num_mmio.is_set()) {
+        res = num_mmio.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.num_interrupts.is_set()) {
+        res = num_interrupts.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.accelerator_state.is_set()) {
+        res = accelerator_state.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.object_id.is_set()) {
+        res = object_id.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.parent.is_set()) {
+        res = parent.update();
+        if (res)
+          goto log_err;
+      }
+      if (p.guid.is_set()) {
+        res = guid.update();
+        if (res)
+          goto log_err;
+      }
+    }
+  }
+  return *this;
+
+log_err:
+  log_.error("operator =") << "property getter failed with (" << res
+                           << ") " << fpgaErrStr(res);
+  throw except(res, OPAECXX_HERE);
+}
+
```

Comment:
Would it make sense to include a `throw_if` in the log?
```C++
log_.throw_if(res, OPAECXX_HERE) << "property getter failed with ...";
```

Reply:
Probably worth thinking about, yes.. but in a separate pull.